internal/verify/test_matrix.go

```diff
@@ -95,31 +105,128 @@ func buildExpectationLookup(table *policyv1.TestTable) (map[testMatrixKey]testMa
 					return nil, fmt.Errorf("found expectations for actions that do not exist in the input actions list: [%s]", strings.Join(extraExpectations, ","))
 				}
 
-				tmExpectation := testMatrixExpectations{actions: expectation.Actions}
-				if n := len(expectation.Outputs); n > 0 {
-					tmExpectation.outputs = make(map[string]*policyv1.Test_OutputEntries, n)
-					for _, oe := range expectation.Outputs {
-						entries := make(map[string]*structpb.Value, len(oe.Expected))
-						for _, entry := range oe.Expected {
-							entries[entry.Src] = entry.Val
-						}
-						tmExpectation.outputs[oe.Action] = &policyv1.Test_OutputEntries{Entries: entries}
-					}
+				lookup[key], err = mergeExpectations(key, lookup[key], expectation.Actions, outputs)
+				if err != nil {
+					return nil, err
 				}
-
-				lookup[key] = tmExpectation
 			}
 		}
 	}
 
 	return lookup, nil
 }
 
-func buildDefaultExpectation(table *policyv1.TestTable) testMatrixExpectations {
+func outputExpectations(expectation *policyv1.TestTable_Expectation) map[string]*policyv1.Test_OutputEntries {
+	outputs := make(map[string]*policyv1.Test_OutputEntries, len(expectation.Outputs))
+	for _, oe := range expectation.Outputs {
+		entries := make(map[string]*structpb.Value, len(oe.Expected))
+		for _, entry := range oe.Expected {
+			entries[entry.Src] = entry.Val
+		}
+		outputs[oe.Action] = &policyv1.Test_OutputEntries{Entries: entries}
+	}
+	return outputs
+}
+
+func mergeExpectations(key testMatrixKey, target *testMatrixExpectations, actions map[string]effectv1.Effect, outputs map[string]*policyv1.Test_OutputEntries) (*testMatrixExpectations, error) {
+	if target == nil {
```

Comment:
@charithe I realised there was a bug in the first iteration of this, where I had

```go
if a == nil {
	return b, nil
}
```

It seems like a nice optimization to avoid allocations but in doing so we end up sharing the same `actions` map between multiple principals; if we have expectations for groups X and Y, where there is some degree of overlap, then the expectations end up being the same for all members of both groups, not just the ones who are actually in both groups.

I've fixed that by making sure we copy the actions into a new map. The same goes for outputs (previously we were copying the `[]*TestTable_OutputExpectations` into a `map[string]*policyv1.Test_OutputEntries` inside the inner loop in `buildExpectationLookup`, but I've moved it to the outer loop now which means we have to avoid the same issue).